### PR TITLE
Initial changes for distributed jmeter executor

### PR DESCRIPTION
**Pull request by hiteshwani29**
## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

**Review by vsukhin — COMMENTED** [review]
Out of 70 files - 62 are copy/paste ones. can we do it better and have common packages for jmeter?

File: contrib/executor/djmeter/pkg/runner/runner.go [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+		if args[i] == "<reportFile>" {
+			args[i] = reportPath
+		}
+
+		if args[i] == "<logFile>" {
+			args[i] = jmeterLogPath
+		}
+	}
+
+	sl, err := slaves.NewClient(execution, r.Params, slaves.SlavesEnvOptions{RunPath: runPath, OverrideJvmArgs: "-Xmn256m -Xms512m -Xmx512m"})
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/runner/runner.go:156]: JVM args shouldn't be hardcoded, we should be able to override them

File: contrib/executor/djmeter/pkg/runner/runner.go [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+	command, args := executor.MergeCommandAndArgs(execution.Command, args)
+	// run JMeter inside repo directory ignore execution error in case of failed test
+	output.PrintLogf("%s Test run command %s %s", ui.IconRocket, command, strings.Join(args, " "))
+	out, err := executor.Run(runPath, command, envManager, args...)
+	if err != nil {
+		return *result.WithErrors(errors.Errorf("jmeter run error: %v", err)), nil
+	}
+	out = envManager.ObfuscateSecrets(out)
+
+	defer sl.DeleteSlaves(slavesNameIpMap)
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/runner/runner.go:197]: I would move it ealier to CreateSlaves

File: contrib/executor/djmeter/pkg/slaves/slave_client.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/kubeshop/testkube/pkg/k8sclient"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	serverPort = 1099
+	localport  = 60001
+)
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/slaves/slave_client.go:21]: should be able to redefine

File: contrib/executor/djmeter/pkg/slaves/slave_client.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	p, err := client.clientSet.CoreV1().Pods(client.namespace).Create(context.Background(), slavePod, metav1.CreateOptions{})
+	if err != nil {
+		errorChan <- err
+		return
+	}
+
+	// Wait for the pod to become ready
+	conditionFunc := isPodReady(context.Background(), client.clientSet, p.Name, client.namespace)
+	timeout := 5 * time.Minute
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/slaves/slave_client.go:96]: move to const at least

File: contrib/executor/djmeter/pkg/slaves/slave_client.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	podNameIpMap := map[string]string{
+		p.Name: p.Status.PodIP,
+	}
+	podIPAddressChan <- podNameIpMap
+}
+
+func (client *Client) DeleteSlaves(slaveNameIpMap map[string]string) {
+	for slaveName := range slaveNameIpMap {
+		output.PrintLog(fmt.Sprintf("Deleting slave %v", slaveName))
+		client.clientSet.CoreV1().Pods(client.namespace).Delete(context.Background(), slaveName, metav1.DeleteOptions{})
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/slaves/slave_client.go:118]: I would record error, if we have it

File: contrib/executor/djmeter/pkg/slaves/slave_utils.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+			Value: envParams.CloudAPIKey,
+		}, {
+			Name:  "RUNNER_CLOUD_API_TLS_INSECURE",
+			Value: fmt.Sprintf("%v", envParams.CloudAPITLSInsecure),
+		}, {
+			Name:  "RUNNER_CLOUD_API_URL",
+			Value: envParams.CloudAPIURL,
+		},
+	}
+}
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/slaves/slave_utils.go:68]: Ideally, it should be in one place in api or we will need to sync them

File: contrib/executor/djmeter/pkg/slaves/slave_utils.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+			},
+			Volumes: []v1.Volume{
+				{
+					Name:         "data-volume",
+					VolumeSource: v1.VolumeSource{EmptyDir: &v1.EmptyDirVolumeSource{}},
+				},
+			},
+		},
+	}
+}
```
Comment (vsukhin) [inline contrib/executor/djmeter/pkg/slaves/slave_utils.go:166]: Should be taken from template, we don't hard code spec

File: contrib/executor/jmeterd/pkg/runner/runner.go [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+	if err != nil {
+		return *result.WithErrors(errors.Errorf("Getting error while creating slaves client: %v", err)), nil
+	}
+
+	slavesNameIpMap, err := sl.CreateSlaves(slavesCount)
+	if err != nil {
+		return *result.WithErrors(errors.Errorf("Getting error while creating slaves nodes: %v", err)), nil
+	}
+
+	args = append(args, fmt.Sprintf("-R %v", GetSlavesIpString(slavesNameIpMap)))
```
Comment (vsukhin) [inline contrib/executor/jmeterd/pkg/runner/runner.go:166]: see comments above

File: contrib/executor/jmeterd/pkg/slaves/slave_client.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	"github.com/kubeshop/testkube/pkg/executor/output"
+	"github.com/kubeshop/testkube/pkg/k8sclient"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	serverPort = 1099
+	localport  = 60001
```
Comment (vsukhin) [inline contrib/executor/jmeterd/pkg/slaves/slave_client.go:20]: again looks like a copy for above file

File: contrib/executor/jmeterd/pkg/slaves/slave_utils.go [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+			Value: envParams.CloudAPIKey,
+		}, {
+			Name:  "RUNNER_CLOUD_API_TLS_INSECURE",
+			Value: fmt.Sprintf("%v", envParams.CloudAPITLSInsecure),
+		}, {
+			Name:  "RUNNER_CLOUD_API_URL",
+			Value: envParams.CloudAPIURL,
+		},
+	}
+}
```
Comment (vsukhin) [inline contrib/executor/jmeterd/pkg/slaves/slave_utils.go:68]: again, one more copy

**Review by vsukhin — COMMENTED** [review]
so, it's good in general, but sounds like only 4 new files, we can organoe it better

**Comment by hiteshwani29** [comment]
Not able to change PR title. Creating a new [PR](https://github.com/kubeshop/testkube/pull/4289) with an appropriate naming convention.
 I'll address these comments in the new PR.
Thanks!

**Pull request closed without merging**